c10/util/IntegerSequence.h

```diff
@@ -0,0 +1,358 @@
+ * so that we can easily port all the algorithms we have available for typelist to std::integer_sequence.
+ */
+template<class T, T value_>
+struct CompileTimeInteger {
+    using IntType = T;
+    static constexpr T value() {
+        return value_;
+    }
+};
+
+/**
+ * Take an integer_sequence/index_sequence and create a typelist using CompileTimeInteger.
+ *
+ * Example:
+ *  iseq_to_typelist_t<std::index_sequence<2, 3>> == typelist<CompileTimeInteger<size_t, 2>, CompileTimeInteger<size_t, 3>>
+ */
+template<class IntegerSequence>
+struct iseq_to_typelist {};
+template<class IntType, IntType... integers>
+struct iseq_to_typelist<std::integer_sequence<IntType, integers...>> {
+    using type = guts::typelist::typelist<CompileTimeInteger<IntType, integers>...>;
+};
+template<class IntegerSequence>
+using iseq_to_typelist_t = typename iseq_to_typelist<IntegerSequence>::type;
+
+
+/**
+ * Take a typelist made from CompileTimeInteger and create an integer_sequence/index_sequence from it.
+ *
+ * Example:
+ *  std::index_sequence<2, 3> == typelist_to_iseq_t<typelist<CompileTimeInteger<size_t, 2>, CompileTimeInteger<size_t, 3>>>
+ */
+template<class TypeList>
+struct typelist_to_iseq {};
+template<class... Types>
+struct typelist_to_iseq<guts::typelist::typelist<Types...>> {
+    using IntType = typename guts::typelist::head_t<guts::typelist::typelist<Types...>>::IntType;
+    static_assert(
+        guts::conjunction<
+            std::is_same<CompileTimeInteger<typename Types::IntType, Types::value()>, Types>...
+        >::value, "All types must be instances of CompileTimeInteger"
+    );
+    static_assert(
+        guts::conjunction<
+            std::is_same<IntType, typename Types::IntType>...
+        >::value, "All types must be based on the same IntType");
+    using type = std::integer_sequence<IntType, Types::value()...>;
+};
+template<>
+struct typelist_to_iseq<guts::typelist::typelist<>> {
+    using type = std::index_sequence<>;
+};
+template<class TypeList>
+using typelist_to_iseq_t = typename typelist_to_iseq<TypeList>::type;
+
+/**
+ * Take the first n elements from an integer_sequence/index_sequence.
+ * Example:
+ *  std::index_sequence<5, 8> == take_t<2, std::index_sequence<5, 8, 3, 5, 6>>
+ */
+template<class ISeq, size_t N>
+using take_t = typelist_to_iseq_t<typelist::take_t<iseq_to_typelist_t<ISeq>, N>>;
+
+/**
+ * Drop the first n elements from an integer_sequence/index_sequence.
+ * Example:
+ *  std::index_sequence<3, 5, 6> == drop_t<2, std::index_sequence<5, 8, 3, 5, 6>>
+ */
+template<class ISeq, size_t N>
+using drop_t = typelist_to_iseq_t<typelist::drop_t<iseq_to_typelist_t<ISeq>, N>>;
+
+/**
+ * Make an index_sequence representing a range of values.
+ * Begin is inclusive, end is exclusive.
+ * Example:
+ *  range_t<3, 8> == std::index_sequence<3, 4, 5, 6, 7>
+ */
+template<size_t begin, size_t end>
+struct range {
```

Comment:
This duplicates some of the functionality in [Metaprogramming.h](https://github.com/pytorch/pytorch/blob/master/c10/util/Metaprogramming.h#L134) - we should probably only have one implementation of this concept. (Also note that the implementation there doesn't require generating the list prefix, which I could imagine mattering in certain use cases.)